components/user-store/org.wso2.carbon.identity.user.store.configuration/src/main/java/org/wso2/carbon/identity/user/store/configuration/utils/UserStoreAttributeMappingParser.java

```diff
@@ -0,0 +1,175 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.user.store.configuration.utils;
+
+import org.apache.axiom.om.OMElement;
+import org.apache.axiom.om.impl.builder.StAXOMBuilder;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.user.store.configuration.model.ChangedUserStoreAttributeDO;
+import org.wso2.carbon.identity.user.store.configuration.model.UserStoreAttributeDO;
+import org.wso2.carbon.utils.CarbonUtils;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import javax.xml.namespace.QName;
+import javax.xml.stream.XMLStreamException;
+
+import static org.wso2.carbon.identity.user.store.configuration.utils.UserStoreConfigurationConstant.ATTRIBUTES_DIR;
+import static org.wso2.carbon.identity.user.store.configuration.utils.UserStoreConfigurationConstant.ATTRIBUTE_ID;
+import static org.wso2.carbon.identity.user.store.configuration.utils.UserStoreConfigurationConstant.CLAIM_URI;
+import static org.wso2.carbon.identity.user.store.configuration.utils.UserStoreConfigurationConstant.DISPLAY_NAME;
+import static org.wso2.carbon.identity.user.store.configuration.utils.UserStoreConfigurationConstant.OPERATION;
+import static org.wso2.carbon.identity.user.store.configuration.utils.UserStoreConfigurationConstant.USERSTORE_DIR;
+import static org.wso2.carbon.identity.user.store.configuration.utils.UserStoreConfigurationConstant.USERSTORE_NAME;
+
+/**
+ * Parser to read and build user store mappings attribute changes.
+ */
+public class UserStoreAttributeMappingParser {
+
+    private static final Map<String, Map<String, ChangedUserStoreAttributeDO>>
+            userStoreAttributeChanges = new HashMap<>();
+    private static final Log LOG = LogFactory.getLog(UserStoreAttributeMappingParser.class);
+
+    private UserStoreAttributeMappingParser() {
+
+        init();
+    }
+
+    private static final class ParserHolder {
+
+        static final UserStoreAttributeMappingParser PARSER = new UserStoreAttributeMappingParser();
+    }
+
+    public static UserStoreAttributeMappingParser getInstance() {
+
+        return ParserHolder.PARSER;
+    }
+
+    private void init() {
+
+        try {
+            for (String fileName : getFileNames()) {
```

Comment:
can getFileNames() become null? if so, handle null check.

Reply:
getFileNames() will not be null.  If no files inside the directory, it will be an empty array. If invalid directory name, throw an IO error.